test/integration/redirect_integration_test.cc

```diff
@@ -171,6 +222,143 @@ TEST_P(RedirectIntegrationTest, BasicInternalRedirect) {
   EXPECT_THAT(waitForAccessLog(access_log_name_, 1), HasSubstr("200 via_upstream -\n"));
 }
 
+TEST_P(RedirectIntegrationTest, InternalRedirectHandlesHttp303) {
+  useAccessLog("%RESPONSE_FLAGS% %RESPONSE_CODE% %RESPONSE_CODE_DETAILS% %RESP(test-header)%");
+  config_helper_.addConfigModifier(
+      [](envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+             hcm) {
+        hcm.set_via("via_value");
+
+        auto* route = hcm.mutable_route_config()->mutable_virtual_hosts(2)->mutable_routes(0);
+        route->mutable_route()
+            ->mutable_internal_redirect_policy()
+            ->mutable_redirect_response_codes()
+            ->Add(303);
+      });
+  initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  default_request_headers_.setHost("handle.internal.redirect");
+  default_request_headers_.setMethod("POST");
+  const std::string& request_body = "foobarbizbaz";
+
+  // First request to original upstream.
+  IntegrationStreamDecoderPtr response =
+      codec_client_->makeRequestWithBody(default_request_headers_, request_body);
+  waitForNextUpstreamRequest();
+  EXPECT_EQ(request_body, upstream_request_->body().toString());
+
+  // Respond with a redirect.
+  redirect_response_.setStatus(303);
+  upstream_request_->encodeHeaders(redirect_response_, true);
+
+  // Second request to redirected upstream.
+  waitForNextUpstreamRequest();
+  EXPECT_EQ("", upstream_request_->body().toString());
+  ASSERT(upstream_request_->headers().EnvoyOriginalUrl() != nullptr);
+  EXPECT_EQ("http://handle.internal.redirect/test/long/url",
+            upstream_request_->headers().getEnvoyOriginalUrlValue());
+  EXPECT_EQ("/new/url", upstream_request_->headers().getPathValue());
+  EXPECT_EQ("authority2", upstream_request_->headers().getHostValue());
+  EXPECT_EQ("via_value", upstream_request_->headers().getViaValue());
+  EXPECT_EQ("GET", upstream_request_->headers().getMethodValue());
```

Comment:
can we add a content length header in the original POST and make sure it's removed here?